# Exception Handling

<!-- TOC -->

- [1. Introduction](#1-introduction)
  - [1.1. Point 1: Exceptions must be Declared](#11-point-1-exceptions-must-be-declared)
  - [1.2. Point 2: Exceptions must be Handled](#12-point-2-exceptions-must-be-handled)
  - [1.3. Point 3: Main logic is separated from the exception handling codes](#13-point-3-main-logic-is-separated-from-the-exception-handling-codes)
- [2. Reference](#2-reference)

<!-- /TOC -->

## 1. Introduction

An exception is an abnormal event that arises during the execution of the program and disrupts the normal flow of the program. Abnormality do occur when your program is running. For example, you might expect the user to enter an integer, but receive a text string; or an unexpected I/O error pops up at runtime. What really matters is "**what happens after an abnormality occurred?**" In other words, "**how the abnormal situations are handled by your program.**" If these exceptions are not handled properly, the program terminates abruptly and may cause severe consequences. For example, the network connections, database connections and files may remain opened; database and file records may be left in an inconsistent state.<sub>感想：现实生活中，出了什么问题，有的人（或许是我自己）只会在问题出现的时候报怨，或者是彰显自己的决定在当时才是睿智的选择，但其实我应该转变自己的思维，去积极的想办法去解决问题。另外，如果自己是犯错误的人，当别人报怨的时候，自己不应当生气，不应该掩饰错误，不应该找各种理由来“减轻”自己的责任，而应该坦诚的承认错误，并接受可能受到的惩罚。别人如果说难听的话，那正好是对自己的心性的考验，应该“闻难听的话，而心里不生反感的情绪”，况且古人说过，闻过则喜。我现在说的挺好，就怕到时候做不到。</sub>

Java has a built-in mechanism for handling runtime errors, referred to as **exception handling**. This is to ensure that you can write robust programs for mission-critical applications.

Older programming languages such as `C` have some drawbacks in exception handing. For example, suppose the programmer wishes to open a file for processing:<sub>我之所以喜欢这篇文章，是因为它提供给我一个视角，这个视角就是将C和Java进行对比，经过“对比”才能够显示Java在处理Exception的时候到底有了哪些改进。总的来说，就是有一种“在编程语言的历史发展过程中，在处理Exception时，不同语言所表现出的手段的变化”</sub>

- The programmers are not made to aware of the exceptional conditions. For example, the file to be opened may not necessarily exist. The programmer therefore did not write codes to test whether the file exists before opening the file.
- Suppose the programmer is aware of the exceptional conditions, he/she might decide to finish the main logic first, and write the exception handling codes later – this "later", unfortunately, usually never happens. In other words, you are not force to write the exception handling codes together with the main logic.
- Suppose the programmer decided to write the exception handling codes, the exception handling codes intertwine with the main logic in many if-else statements. This makes main logic hard to follow and the entire program hard to read. For example,

这三段总结：

- 第一，就是：有没有意识到问题的存在？
- 第二，就是：如果意识到问题的存在，那么要不要进行处理？
- 第三，就是：如果要进行处理，那么采取什么样的方法进行处理呢？

换种方式说：

- 第一点呢，是属于“认知”或“意识”方面的层面，它指出的是一个“存在性”的问题，就是能不能感知到一件事物究竟存在或不存在？
- 第二点呢，是属于“态度”层面。问题存在了，如果持一个“负责认真的态度”，就应该对它进行处理；如果持一个“不负责任的态度”，那可能就不对它进行处理。
- 第三点呢，是属于“手段”层面。就是怎么去处理这个问题，采用什么样的方式方法。

```c
if (file exists) {
   open file;
   while (there is more records to be processed) {
      if (no IO errors) {
         process the file record
      } else {
         handle the errors
      }
   }
   if (file is opened) close the file;
} else {
   report the file does not exist;
}
```

Java overcomes these drawbacks by building the **exception handling** into the language rather than leaving it to the discretion of the programmers:

- You will be informed of the exceptional conditions that may arise in calling a method - Exceptions are declared in the method's signature.
- You are forced to handle exceptions while writing the main logic and cannot leave them as an afterthought - Your program cannot compiled without the exception handling codes.
- Exception handling codes are separated from the main logic - Via the try-catch-finally construct.

这三段总结：

- 第一，在“认知”层面，不再依赖于programmer的认知水平，而是由method signature来告诉programmer是否存在问题。
- 第二，在“态度”层面，不再依赖于programmer的“态度”或“当时的心情”，而是由Java Compiler来“强制”提醒programmer。
- 第三，在“手段”层面，Java提供了try-catch-finally的方式来处理，这就尽量避免了main logic和exception handling的代码相互“交织”的问题。

Let's look into these three points in more details.

### 1.1. Point 1: Exceptions must be Declared

As an example, suppose that you want to use a `java.util.Scanner` to perform formatted input from a disk file. The signature of the Scanner's constructor with a `File` argument is given as follows:

```java
public Scanner(File source) throws FileNotFoundException;
```

The **method's signature** informs the programmers that an exceptional condition "file not found" may arise. By declaring the exceptions in the method's signature, programmers are made to aware of the exceptional conditions in using the method.

### 1.2. Point 2: Exceptions must be Handled

If a method declares an exception in its signature, you cannot use this method without handling the exception - you can't compile the program.

Example 1: The program did not handle the exception declared, resutled in compilation error.

```java
import java.util.Scanner;
import java.io.File;
public class ScannerFromFile {
   public static void main(String[] args) {
      Scanner in = new Scanner(new File("test.in"));
      // do something ...
   }
}
```

```txt
ScannerFromFile.java:5: unreported exception java.io.FileNotFoundException; must be caught or declared to be thrown
      Scanner in = new Scanner(new File("test.in"));
                   ^
```

To use a method that declares an exception in its signature, you MUST either:

- provide exception handling codes in a "try-catch" or "try-catch-finally" construct, or
- not handling the exception in the current method, but declare the exception to be thrown up the call stack for the next higher-level method to handle.

Example 2: Catch the exception via a "try-catch" (or "try-catch-finally") construct.

```java
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
public class ScannerFromFileWithCatch {
   public static void main(String[] args) {
      try {
         Scanner in = new Scanner(new File("test.in"));
         // do something if no exception ...
         // you main logic here in the try-block
      } catch (FileNotFoundException ex) { // error handling separated from the main logic
         ex.printStackTrace();             // print the stack trace
      }
   }
}
```

If the file cannot be found, the exception is caught in the catch-block. In this example, the error handler simply prints the stack trace, which provides useful information for debugging. In some situations, you may need to perform some clean-up operations, or open another file instead. Take note that the main logic in the try-block is separated from the error handling codes in the catch-block.

Example 3: You decided not to handle the exception in the current method, but throw the exception up the call stack for the next higher-level method to handle.

```java
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
public class ScannerFromFileWithThrow {
   public static void main(String[] args) throws FileNotFoundException {  
                                 // to be handled by next higher-level method
      Scanner in = new Scanner(new File("test.in"));
                                // this method may throw FileNotFoundException
      // main logic here ...
   }
}
```

In this example, you decided not to handle the `FileNotFoundException` thrown by the `Scanner(File)` method (with try-catch). Instead, the caller of `Scanner(File)` - the `main()` method - declares in its signature "`throws FileNotFoundException`", which means that this exception will be thrown up the call stack, for the next higher-level method to handle. In this case, the next higher-level method of `main()` is the JVM, which simply terminates the program and prints the stack trace.

### 1.3. Point 3: Main logic is separated from the exception handling codes

As shown in Example 2, the **main logic** is contained in the **try-block**, while the **exception handling codes** are kept in the **catch-block**(s) separated from the **main logic**. This greatly improves the readability of the program.

For example, a Java program for file processing could be as follows:

```java
try {
   // Main logic here
   open file;
   process file;
   ......
} catch (FileNotFoundException ex) {   // Exception handlers below
   // Exception handler for "file not found"
} catch (IOException ex) {
   // Exception handler for "IO errors"
} finally {
  close file;      // always try to close the file
}
```

## 2. Reference

- [Java Programming
Exception Handling & Assertion](https://www.ntu.edu.sg/home/ehchua/programming/java/J5a_ExceptionAssert.html)
